include/deal.II/base/bounding_box.h

```diff
@@ -0,0 +1,112 @@
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_base_bounding_box_h
+#define dealii_base_bounding_box_h
+
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/point.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A class that represents a bounding box in a space with arbitrary dimension
+ * <tt>spacedim</tt>.
+ *
+ * Objects of this class are used to represent bounding boxes. They are,
+ * among other uses, useful in parallel distributed meshes to give a general
+ * description the owners of each portion of the mesh.
+ *
+ * Boundig boxes are represented by two vertices (bottom left and top right).
+ * Geometrically, are represent the following:
+ * 1 d: a segment (represented by its vertices)
+ * 2 d: a rectangle (represented by two vertices V opposed to the center)
+ * .--------V
+ * |        |
+ * V--------.
+ * 3 d: a cuboid (represented by two vertices V opposed to the center
+ *               and not owned by the same face)
+ *   .------V
+ *  /      /|
+ * .------. |
+ * |      | /
+ * |      |/
+ * V------.
+ *
+ * Notice the sides are always parallel to the respective axis.
+ *
+ * @author Giovanni Alzetta, 2017
+ */
+template <int spacedim, typename Number=double>
+class BoundingBox
+{
+public:
+  /**
+   * Standard constructor. Creates an object that corresponds to a empty box,
+   * i.e. an degenerate box with both points being the origin.
+   */
+  BoundingBox ();
+
+  /**
+   * Standard constructor for non-empty boxes: it uses a pair of points
+   * which describe the box: one for the bottom and one for the top
+   * corner.
+   */
+  BoundingBox (const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &boundary_points);
+
+  /**
+   * Returns the boundary_points
+   */
+  const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &get_boundary_points () const;
+
+  /**
+   * Returns true if the point is inside the Bounding Box, false otherwise
+   */
+  bool point_inside (const Point<spacedim, Number> &p) const;
+
+private:
+  std::pair<Point<spacedim, Number>,Point<spacedim,Number>> boundary_points;
+};
+
+/*------------------------------- Inline functions: Point ---------------------------*/
+
+#ifndef DOXYGEN
+
+// At least clang-3.7 requires us to have a user-defined constructor
+// and we can't use 'BoundingBox<int spacedim>::BoundingBox ()' here.
+template <int spacedim, typename Number >
+inline
+BoundingBox<spacedim, Number>::BoundingBox ()
+{
+}
+
+
+template <int spacedim, typename Number>
+inline
+BoundingBox<spacedim, Number>::BoundingBox (const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &boundary_points)
+{
+  for (unsigned int i=0; i<spacedim; ++i)
```

Comment:
i wonder what happens in `release` mode here as `Assert` is gone... @bangerth probably knows :smile: but one could do 
```
#ifdef DEBUG
   for ()

#endif
```
to be sure there are no warnings/etc.

Reply:
No, we do this all the time without the `#ifdef`. In release mode, the compiler just sees
```
  for (unsigned int i=.....)
    ;
```
which it duly optimizes away completely because nothing in the loop header has any side effects. 

So this is fine.